### PR TITLE
Remove usage of controllerImageVersion values field

### LINKED ISSUE #7859: cli: cannot override install version
### What is the issue?

Overriding the image version for control plane components does not work when done through `linkerd install --set "linkerdVersion=<version>"`

```
:; bin/linkerd install --set linkerdVersion=edge-22.2.1 | rg 'edge-22.2.1'
W0210 17:21:52.122688  345990 warnings.go:70] discovery.k8s.io/v1beta1 EndpointSlice is deprecated in v1.21+, unavailable in v1.25+; use discovery.k8s.io/v1 EndpointSlice
    linkerdVersion: edge-22.2.1
        image: cr.l5d.io/linkerd/policy-controller:edge-22.2.1
```

The only image that seems to be overidden is the policy-controller. The policy-controller may override the cli version using `policyController.image.version`, whereas the controllers can be overidden independently of other components through `controllerImageVersion`.

```
image: {{.Values.controllerImage}}:{{default .Values.linkerdVersion .Values.controllerImageVersion}} vs
image: {{.Values.policyController.image.name}}:{{.Values.policyController.image.version | default .Values.linkerdVersion}}
```

`controllerImageVersion` has been removed and is no longer supported. We should instead just use the global `linkerdVersion` override.

### How can it be reproduced?

Install linkerd with a recent edge version and try to override the image versions by setting `linkerdVersion`

### Logs, error output, etc

```
W0210 16:35:40.137514  296270 warnings.go:70] discovery.k8s.io/v1beta1 EndpointSlice is deprecated in v1.21+, unavailable in v1.25+; use discovery.k8s.io/v1 EndpointSlice
    linkerdVersion: edge-22.2.1
        image: cr.l5d.io/linkerd/policy-controller:edge-22.2.1
```

### output of `linkerd check -o short`

N/A

### Environment

linkerd version: edge-22.5.1

### Possible solution

`controllerImageVersion` can no longer be set in the `values.yaml` file, but it is still used in code:

https://github.com/linkerd/linkerd2/blob/68b63269d952b05cc721581dfa4672ad2e775964/pkg/charts/linkerd2/values.go#L243

We should remove it from our code so that the cli does not pass any default values to it. Currently, the controller version gets initialised, it is never empty, so we never use the `linkerdVersion` default in the template. See code above.

N.B helm works just fine to override the image version

### Additional context

_No response_

### Would you like to work on fixing this bug?

yes

### PR DESCRIPTION
**Pull request by mateiidavid**
We have removed the option of configuring the controller image version
through a dedicated CLI flag (no more `--controller-image-version`). All
of our controller images will now mirror `linkerdVersionValue`, and that
should be the only way of overwriting the images in manifests. The
exception here is the policy controller which seems to have its own
field (no corresponding flag, it still has to be set through `--set`).

This change removes the unused `controllerImageVersion` field, first
from the tests, and then from the actual chart values structure. Note
that at this point in time, it is impossible to use
`--controller-image-version` through Helm, yet it still seems to be
working for the CLI.

As part of this change:

* We configure the charts to use `linkerdVersionValue` instead of
  `controlPlaneImageVersion` (or default to it where appropriate).
* We add the stringslicevar flag (i.e `--set`) to the flagset we use in
  upgrade tests. This means instead of testing value overrides through a
  dedicated flag, we can now make use of `--set` in upgrade tests. We
  first set the linkerdVersionValue in the install option and then
  override the policy controller image version and the linkerd
  controller image version to test flags work as expected.
* We remove hardcoded values from healthcheck test.
* We remove field from chart values struct.

Closes #7859

Signed-off-by: Matei David <matei@buoyant.io>

**Note to reviewers**: I _may_ have made some mistakes with the test manifests. I ran `go test ./... --update` quite a few times so it'd be good to double check the golden files are in an ok shape.

### CODE REVIEW

File: charts/linkerd-control-plane/templates/identity.yaml [hunk 1]
```diff
@@ -155,7 +155,7 @@ spec:
         env:
         - name: LINKERD_DISABLED
           value: "linkerd-await cannot block the identity controller"
-        image: {{.Values.controllerImage}}:{{default .Values.linkerdVersion .Values.controllerImageVersion}}
+        image: {{.Values.controllerImage}}:{.Values.linkerdVersion}}
```
Comment (adleong) [inline charts/linkerd-control-plane/templates/identity.yaml:158]: Missing { in the substitution here

**Review by adleong — COMMENTED** [review]
I'm in huge favor of simplifying our Helm values :100:

It's a bit confusing that we have different methods for setting the proxy and controller image versions.  i.e. if I want to install Linkerd at a different version from my CLI, I would run something like:

```
linkerd install --set linkerdVersion=edge-x.y.z --proxy-version edge-x.y.z
```

What do you think about having a CLI flag (--linkerd-version?) which sets both Values.LinkerdVersion and Values.Proxy.Image.Version?  The --set flag can then be used for more fine-grained control.

e.g.
to use a specific version for both controller and proxy (this is the case that I use most often):
```
linkerd install --linkerd-version edge-x.y.z
```
to override the proxy version
```
linkerd install --linkerd-version edge-x.y.z --set proxy.image.version=my-cool-proxy
```

**Comment by olix0r** [comment]
> What do you think about having a CLI flag (--linkerd-version?) which sets both Values.LinkerdVersion and Values.Proxy.Image.Version?

Note that it's not really a good idea to have a given CLI version install other linkerd versions unless you're really truly sure that the templates are otherwise exactly the same. I'd prefer to remove this CLI flags entirely and force use of the helm values if they need to be set.

File: charts/linkerd-control-plane/templates/identity.yaml (hunk 1 above)
Comment (mateiidavid) [reply charts/linkerd-control-plane/templates/identity.yaml]: Great spot!

**Comment by mateiidavid** [comment]
@adleong that's a good suggestion! I think we might be able to do that through Helm template values -- I actually thought it was possible to do already --  in the proxy partial, we seem to default to the `linkerdVersion` value:

https://github.com/linkerd/linkerd2/blob/c5d36f26577392d25eb9c176607f683bc34ade2f/charts/partials/templates/_proxy.tpl#L131

This value doesn't seem to take though. What I mean by that is for some reason, even if we specify `linkerdVersion`, the proxy partial won't inherit it (unless I'm missing something very obvious):

```
:; bin/linkerd install --ignore-cluster --set "linkerdVersion=xyz" | rg 'cr.l5d.io/linkerd/proxy'
        name: cr.l5d.io/linkerd/proxy
        name: cr.l5d.io/linkerd/proxy-init
        image: cr.l5d.io/linkerd/proxy:git-ca6c61aa
        image: cr.l5d.io/linkerd/proxy-init:v1.5.2
        image: cr.l5d.io/linkerd/proxy:git-ca6c61aa
        image: cr.l5d.io/linkerd/proxy-init:v1.5.2
        image: cr.l5d.io/linkerd/proxy:git-ca6c61aa
        image: cr.l5d.io/linkerd/proxy-init:v1.5.2

```

@olix0r 

> I'd prefer to remove this CLI flags entirely and force use of the helm values if they need to be set.

We could move away from it. I think the biggest challenge is getting all of the unit tests to pass after we deprecate it, that should be a bit easier now through this change though. An alternative would be to just hide the option in edge and stable releases, like we've done with `--controller-image-version`. I'm easy either way.

**Review by olix0r — APPROVED** [review]
lgtm

**Review by alpeb — COMMENTED** [review]
The change makes sense to me :+1: 
I found another appearance of `controllerImageVersion` in `cli/cmd/install_helm_test.go`, which seems to propagate into the associated helm test golden files as well; would you remove that as well?

**Review by kleimkuhler — APPROVED** [review]
Looks good modulo alpeb's comment about not passing in an empty `templateOpts`.

**Review by alpeb — APPROVED** [review]
>  I added a wrapper type to hold the flag, flagSet and our template overrides.

Nice refactor! :shipit:

**Pull request merged**